# Regressão Logística I
## Tarefa II

Vamos trabalhar com a mesma base do exercício anterior, mas vamos aprofundar um pouco mais a nossa regressão.

In [16]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

import statsmodels.formula.api as smf

ModuleNotFoundError: No module named 'statsmodels'

In [3]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data'

df = pd.read_csv(url, 
                 names=['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg',
                        'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'])
df['flag_doente'] = (df['num']!=0).astype('int64')
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak   
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3  \
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

A descrição das variáveis está recortada abaixo:
- age: idade do paciente em anos
- sex: sexo (1 = male; 0 = female)  
- cp: tipo de dor no peito
  - 1: angina típica
  - 2: angina atípica
  - 3: dor não-angina
  - 4: assintomático
- trestbps: pressão sanguínea em repouso (em mm Hg na admissão ao hospital
- chol: colesterol sérico em mg/dl
- fbs: (açúcar no sangue em jejum > 120 mg/dl) (1 = True; 0 = False)
- restecg: resultados eletrocardiográficos em repouso
  - 0: normal
  - 1: tendo anormalidade da onda ST-T (Inversões de onda T e / ou ST com elevação ou depressão de > 0.05 mV)
  - 2: mostrando hipertrofia ventricular esquerda provável ou definitiva pelos critérios de Estes
- thalach: frequência cardíaca máxima alcançada
- exang: angina induzida por exercício(1 = sim; 0 = não)
- oldpeak = Depressão de ST induzida por exercício em relação ao repouso
- slope: Depressão de ST induzida por exercício em relação ao repouso
  - 1: inclinação ascendente
  - 2: estável
  - 3: inclinação descendente
- ca: número de vasos principais (0-3) coloridos por fluorosopia
- thal: 3 = normal; 6 = defeito corrigido; 7 = defeito reversível
- num: diagnóstico de doença cardíaga (status de doença angiográfica)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   age          303 non-null    float64
 1   sex          303 non-null    float64
 2   cp           303 non-null    float64
 3   trestbps     303 non-null    float64
 4   chol         303 non-null    float64
 5   fbs          303 non-null    float64
 6   restecg      303 non-null    float64
 7   thalach      303 non-null    float64
 8   exang        303 non-null    float64
 9   oldpeak      303 non-null    float64
 10  slope        303 non-null    float64
 11  ca           303 non-null    object 
 12  thal         303 non-null    object 
 13  num          303 non-null    int64  
 14  flag_doente  303 non-null    int64  
dtypes: float64(11), int64(2), object(2)
memory usage: 35.6+ KB


1. Considere o script que monta a análise bivariada que você fez na tarefa anterior. Transforme esse script em uma função, que deve:
- Ter como parâmetros de entrada:
    - Um *dataframe* contendo os dados a serem avaliados
    - Um *string* contendo o nome da variável resposta
    - Um *string* contendo o nome da variável explicativa
- E deve retornar um *dataframe* com os dados da bivariada. 
**Monte** a mesma bivariada pelo menos três variáveis qualitativas do *data-frame*. Qual delas parece discriminar mais o risco?

In [5]:
def tab_func(df, resposta, explicativa):
    tab = pd.crosstab(df[explicativa], df[resposta], margins=True)
    tab['media'] = tab[1] / tab['All']
    tab['odds'] = tab['media'] / (1 - tab['media'])
    tab['odds_ratio'] = tab['odds'] / tab.loc['All', 'odds']
    tab['logito'] = np.log(tab['odds'])
    tab['woe'] = np.log(tab['odds_ratio'])
    return tab

In [6]:
tab_sex = tab_func(df=df, resposta='flag_doente', explicativa='sex')
tab_sex


flag_doente    0    1  All     media      odds  odds_ratio    logito       woe
sex                                                                           
0.0           72   25   97  0.257732  0.347222    0.409672 -1.057790 -0.892398
1.0           92  114  206  0.553398  1.239130    1.461996  0.214410  0.379802
All          164  139  303  0.458746  0.847561    1.000000 -0.165392  0.000000

In [7]:
tab_fbs = tab_func(df=df, resposta='flag_doente', explicativa='fbs')
tab_fbs

flag_doente    0    1  All     media      odds  odds_ratio    logito       woe
fbs                                                                           
0.0          141  117  258  0.453488  0.829787    0.979030 -0.186586 -0.021193
1.0           23   22   45  0.488889  0.956522    1.128558 -0.044452  0.120941
All          164  139  303  0.458746  0.847561    1.000000 -0.165392  0.000000

In [8]:
tab_exang = tab_func(df=df, resposta='flag_doente', explicativa='exang')
tab_exang

flag_doente    0    1  All     media      odds  odds_ratio    logito       woe
exang                                                                         
0.0          141   63  204  0.308824  0.446809    0.527170 -0.805625 -0.640233
1.0           23   76   99  0.767677  3.304348    3.898655  1.195239  1.360632
All          164  139  303  0.458746  0.847561    1.000000 -0.165392  0.000000

2. Monte uma função semelhante para categorizar variáveis quantitativas contínuas (com muitas categorias) como ```age```.  
    Além dos mesmos parâmetros da função anterior, defina mais um parâmetro como número de categorias que você deseja quebrar. Defina um valor '*default*' de 5 grupos para este parâmetro.  

In [9]:
def tab_func_2(df, resposta, explicativa, num_categorias=5):
    df['grupo'] = pd.qcut(df[explicativa], num_categorias, duplicates='drop')
    tab = pd.crosstab(df['grupo'], df[resposta], margins=True)
    tab['media'] = tab[1] / tab['All']
    tab['odds'] = tab['media'] / (1 - tab['media'])
    tab['odds_ratio'] = tab['odds'] / tab.loc['All', 'odds']
    tab['logito'] = np.log(tab['odds'])
    tab['woe'] = np.log(tab['odds_ratio'])
    return tab

In [10]:
tab_age = tab_func_2(df=df, resposta='flag_doente', explicativa='age')
tab_age

flag_doente       0    1  All     media      odds  odds_ratio    logito   
grupo                                                                     
(28.999, 45.0]   47   16   63  0.253968  0.340426    0.401653 -1.077559  \
(45.0, 53.0]     42   22   64  0.343750  0.523810    0.618020 -0.646627   
(53.0, 58.0]     32   39   71  0.549296  1.218750    1.437950  0.197826   
(58.0, 62.0]     13   32   45  0.711111  2.461538    2.904261  0.900787   
(62.0, 77.0]     30   30   60  0.500000  1.000000    1.179856  0.000000   
All             164  139  303  0.458746  0.847561    1.000000 -0.165392   

flag_doente          woe  
grupo                     
(28.999, 45.0] -0.912166  
(45.0, 53.0]   -0.481235  
(53.0, 58.0]    0.363218  
(58.0, 62.0]    1.066179  
(62.0, 77.0]    0.165392  
All             0.000000

In [11]:
tab_trestbps = tab_func_2(df=df, resposta='flag_doente', explicativa='trestbps')
tab_trestbps

flag_doente        0    1  All     media      odds  odds_ratio    logito   
grupo                                                                      
(93.999, 120.0]   60   37   97  0.381443  0.616667    0.727578 -0.483427  \
(120.0, 126.0]    10   15   25  0.600000  1.500000    1.769784  0.405465   
(126.0, 134.0]    35   27   62  0.435484  0.771429    0.910175 -0.259511   
(134.0, 144.6]    34   24   58  0.413793  0.705882    0.832840 -0.348307   
(144.6, 200.0]    25   36   61  0.590164  1.440000    1.698993  0.364643   
All              164  139  303  0.458746  0.847561    1.000000 -0.165392   

flag_doente           woe  
grupo                      
(93.999, 120.0] -0.318034  
(120.0, 126.0]   0.570858  
(126.0, 134.0]  -0.094119  
(134.0, 144.6]  -0.182914  
(144.6, 200.0]   0.530036  
All              0.000000

3. Construa um modelo de regressão logística com as variáveis qualitativas: ```sex + cp +  trestbps``` e com a variável quantitativa ```age```.

**Interprete os parâmetros.**

In [12]:
reglog = smf.logit('flag_doente ~ C(sex) + C(cp, Treatment(4)) + trestbps + age', data=df).fit()
reglog.summary()

NameError: name 'smf' is not defined

4. Avalie o seu modelo quanto a **calibragem**:
- Calcule a probabilidade de evento predita segundo o seu modelo
- Categorize essa probabilidade em G=5 grupos
- Calcule a probabilidade de evento predita média por grupo
- Calcule a taxa de eventos (média da variável indicadora de eventos) por grupo
- Compare graficamente o valor eperado versus observado para a taxa de maus por grupo

In [14]:
df['predito'] = reglog.predict(df)
df.head()

NameError: name 'reglog' is not defined

In [15]:
cat_pred = pd.qcut(df['predito'], 5, duplicates='drop')
cat_pred.value_counts().sort_index()

KeyError: 'predito'

5. Avalie o seu modelo quanto a discriminação calculando acurácia, GINI e KS.

6. tente melhorar o modelo obtido, por exemplo inserindo ou removendo variáveis.  
    Avalie as características do seu modelo (calibragem e acurácia).